vercel/resource_firewall_config.go

```diff
@@ -0,0 +1,878 @@
+	Action         Mitigate         `tfsdk:"action"`
+}
+
+func (r *FirewallRule) Conditions() []client.ConditionGroup {
+	var groups []client.ConditionGroup
+	for _, group := range r.ConditionGroup {
+		var conditions []client.Condition
+		for _, condition := range group.Conditions {
+			conditions = append(conditions, client.Condition{
+				Type:  condition.Type.ValueString(),
+				Op:    condition.Op.ValueString(),
+				Neg:   condition.Neg.ValueBool(),
+				Key:   condition.Key.ValueString(),
+				Value: condition.Value.ValueString(),
+			})
+		}
+		groups = append(groups, client.ConditionGroup{
+			Conditions: conditions,
+		})
+	}
+	return groups
+}
+
+func (r *FirewallRule) Mitigate() client.Mitigate {
+	mit := client.Mitigate{
+		Action: r.Action.Action.ValueString(),
+	}
+	if r.Action.RateLimit != nil {
+		keys := make([]string, len(r.Action.RateLimit.Keys))
+		for i, k := range r.Action.RateLimit.Keys {
+			keys[i] = k.ValueString()
+		}
+		mit.RateLimit = &client.RateLimit{
+			Algo:   r.Action.RateLimit.Algo.ValueString(),
+			Window: r.Action.RateLimit.Window.ValueInt64(),
+			Limit:  r.Action.RateLimit.Limit.ValueInt64(),
+			Keys:   keys,
+			Action: r.Action.RateLimit.Action.ValueString(),
+		}
+	}
+	if r.Action.Redirect != nil {
+		mit.Redirect = &client.Redirect{
+			Location:  r.Action.Redirect.Location.ValueString(),
+			Permanent: r.Action.Redirect.Permanent.ValueBool(),
+		}
+	}
+	if !r.Action.ActionDuration.IsNull() {
+		mit.ActionDuration = r.Action.ActionDuration.ValueString()
+	}
+	return mit
+}
+
+func fromFirewallRule(rule client.FirewallRule, ref FirewallRule) FirewallRule {
+	r := FirewallRule{
+		ID:          types.StringValue(rule.ID),
+		Name:        types.StringValue(rule.Name),
+		Description: types.StringValue(rule.Description),
+		Active:      types.BoolValue(rule.Active),
+	}
+	if rule.Active && ref.Active == types.BoolNull() {
+		r.Active = ref.Active
+	}
+
+	r.Action = fromMitigate(rule.Action.Mitigate, ref.Action)
+	var conditionGroups = make([]ConditionGroup, len(rule.ConditionGroup))
+	for j, group := range rule.ConditionGroup {
+		var conditions = make([]Condition, len(group.Conditions))
+		for k, condition := range group.Conditions {
+			var cond = Condition{}
+			if len(ref.ConditionGroup) > j && len(ref.ConditionGroup[j].Conditions) > k {
+				cond = ref.ConditionGroup[j].Conditions[k]
+			}
+			conditions[k] = fromCondition(condition, cond)
+		}
+		conditionGroups[j] = ConditionGroup{
+			Conditions: conditions,
+		}
+	}
+	r.ConditionGroup = conditionGroups
+	// Description and active can be optional
+	if rule.Description == "" && ref.Description == types.StringNull() {
+		r.Description = ref.Description
+	}
+	if rule.Active && ref.Active == types.BoolNull() {
+		r.Active = ref.Active
+	}
+
+	return r
+}
+
+type Mitigate struct {
+	Action         types.String `tfsdk:"action"`
+	RateLimit      *RateLimit   `tfsdk:"rate_limit"`
+	Redirect       *Redirect    `tfsdk:"redirect"`
```

Comment:
Could consider using `types.Object` rather than `*Redirect` for these kinda fields. Not a biggie if it makes the code a lot messier though.